epochStart/metachain/rewards.go

```diff
@@ -230,101 +245,141 @@ func (r *rewardsCreator) VerifyRewardsMiniBlocks(metaBlock *block.MetaBlock, val
 }
 
 // CreateMarshalizedData creates the marshalized data to be sent to shards
-func (r *rewardsCreator) CreateMarshalizedData(body *block.Body) map[string][][]byte {
+func (rc *rewardsCreator) CreateMarshalizedData(body *block.Body) map[string][][]byte {
+	if check.IfNil(body) {
+		return nil
+	}
+
 	txs := make(map[string][][]byte)
 
 	for _, miniBlock := range body.MiniBlocks {
 		if miniBlock.Type != block.RewardsBlock {
 			continue
 		}
 
-		broadCastTopic := createBroadcastTopic(r.shardCoordinator, miniBlock.ReceiverShardID)
+		broadCastTopic := createBroadcastTopic(rc.shardCoordinator, miniBlock.ReceiverShardID)
 		if _, ok := txs[broadCastTopic]; !ok {
 			txs[broadCastTopic] = make([][]byte, 0, len(miniBlock.TxHashes))
 		}
 
 		for _, txHash := range miniBlock.TxHashes {
-			rwdTx, err := r.currTxs.GetTx(txHash)
+			rwdTx, err := rc.currTxs.GetTx(txHash)
 			if err != nil {
 				continue
 			}
 
-			marshaledData, err := r.marshalizer.Marshal(rwdTx)
+			marshalizedData, err := rc.marshalizer.Marshal(rwdTx)
 			if err != nil {
 				continue
 			}
 
-			txs[broadCastTopic] = append(txs[broadCastTopic], marshaledData)
+			txs[broadCastTopic] = append(txs[broadCastTopic], marshalizedData)
 		}
 	}
 
 	return txs
 }
 
 // SaveTxBlockToStorage saves created data to storage
-func (r *rewardsCreator) SaveTxBlockToStorage(metaBlock *block.MetaBlock, body *block.Body) {
+func (rc *rewardsCreator) SaveTxBlockToStorage(_ *block.MetaBlock, body *block.Body) {
+	if check.IfNil(body) {
+		return
+	}
+
 	for _, miniBlock := range body.MiniBlocks {
 		if miniBlock.Type != block.RewardsBlock {
 			continue
 		}
 
 		for _, txHash := range miniBlock.TxHashes {
-			rwdTx, err := r.currTxs.GetTx(txHash)
+			rwdTx, err := rc.currTxs.GetTx(txHash)
 			if err != nil {
 				continue
 			}
 
-			marshaledData, err := r.marshalizer.Marshal(rwdTx)
+			marshalizedData, err := rc.marshalizer.Marshal(rwdTx)
 			if err != nil {
 				continue
 			}
 
-			_ = r.rewardsStorage.Put(txHash, marshaledData)
+			_ = rc.rewardsStorage.Put(txHash, marshalizedData)
 		}
 
-		for _, mbHeader := range metaBlock.MiniBlockHeaders {
-			if mbHeader.Type != block.RewardsBlock {
-				continue
-			}
-			if mbHeader.ReceiverShardID != miniBlock.ReceiverShardID {
-				continue
-			}
-
-			marshaledData, err := r.marshalizer.Marshal(miniBlock)
-			if err != nil {
-				continue
-			}
-
-			_ = r.miniBlockStorage.Put(mbHeader.Hash, marshaledData)
+		marshalizedData, err := rc.marshalizer.Marshal(miniBlock)
```

Comment:
the previous for was used also to test that the current miniblock has a correspondent in the miniblockHeaders and it was contained in the metablock header. Is it not needed?

Reply:
Actually these checks are done when creating the metablock, when receiving and processing it.I have refactored this part after a comment / suggestion of @sasurobert above

Reply:
ok